extensions/replication/queueProcessor/MultipleBackendTask.js

```diff
@@ -0,0 +1,175 @@
+const jsutil = require('arsenal').jsutil;
+
+const QueueProcessorTask = require('./QueueProcessorTask');
+const attachReqUids = require('../utils/attachReqUids');
+
+class MultipleBackendTask extends QueueProcessorTask {
+
+    _setupRolesOnce(entry, log, cb) {
+        log.debug('getting bucket replication', { entry: entry.getLogInfo() });
+        const entryRolesString = entry.getReplicationRoles();
+        let entryRoles;
+        if (entryRolesString !== undefined) {
+            entryRoles = entryRolesString.split(',');
+        }
+        if (entryRoles === undefined || entryRoles.length !== 1) {
+            log.error('expecting a single role in entry replication ' +
+            'configuration when replicating to an external location', {
+                method: 'MultipleBackendTask._setupRolesOnce',
+                entry: entry.getLogInfo(),
+                roles: entryRolesString,
+            });
+            return cb(errors.BadRole);
+        }
+        this.sourceRole = entryRoles[0];
+
+        this._setupSourceClients(this.sourceRole, log);
+
+        const req = this.S3source.getBucketReplication({
+            Bucket: entry.getBucket(),
+        });
+        attachReqUids(req, log);
+        return req.send((err, data) => {
+            if (err) {
+                log.error('error getting replication configuration from S3', {
+                    method: 'MultipleBackendTask._setupRolesOnce',
+                    entry: entry.getLogInfo(),
+                    origin: 'source',
+                    peer: this.sourceConfig.s3,
+                    error: err.message,
+                    httpStatus: err.statusCode,
+                });
+                // eslint-disable-next-line no-param-reassign
+                err.origin = 'source';
+                return cb(err);
+            }
+            const replicationEnabled = data.ReplicationConfiguration.Rules
+                .some(rule => rule.Status === 'Enabled' &&
+                    entry.getObjectKey().startsWith(rule.Prefix));
+            if (!replicationEnabled) {
+                log.debug('replication disabled for object', {
+                    method: 'MultipleBackendTask._setupRolesOnce',
+                    entry: entry.getLogInfo(),
+                });
+                return cb(errors.PreconditionFailed.customizeDescription(
+                    'replication disabled for object'));
+            }
+            const roles = data.ReplicationConfiguration.Role.split(',');
+            if (roles.length !== 1) {
+                log.error('expecting a single role in bucket replication ' +
+                'configuration when replicating to an external location', {
+                    method: 'MultipleBackendTask._setupRolesOnce',
+                    entry: entry.getLogInfo(),
+                    roles,
+                });
+                return cb(errors.BadRole);
+            }
+            if (roles[0] !== entryRoles[0]) {
```

Comment:
I know that we have implemented the same before, but I am conflicted on this one. For roles, if they are updated in the bucket's configuration we should just use the latest one. There is no need to penalize the user for switching the roles, especially since the crr backlog is async.